File: src/lib/categorize.test.ts
import { describe, it, expect } from 'vitest';
import { suggestCategory } from './categorize';
import type { CategoryRule } from '../db/schema';

function rule(p: Partial<CategoryRule> & { keyword: string; categoryId: string }): CategoryRule {
  return { priority: 0, ...p };
}

describe('suggestCategory', () => {
  it('returns null when no rule matches', () => {
    expect(suggestCategory('Uber para o hotel', [])).toBeNull();
  });

  it('matches a keyword as a case-insensitive substring', () => {
    const rules = [rule({ keyword: 'uber', categoryId: 'cat-transport' })];
    expect(suggestCategory('UBER *TRIP', rules)).toBe('cat-transport');
  });

  it('picks the highest-priority match when two rules match', () => {
    const rules = [
      rule({ keyword: 'uber', categoryId: 'cat-transport', priority: 1 }),
      rule({ keyword: 'eats', categoryId: 'cat-food', priority: 5 }),
    ];
    expect(suggestCategory('UBER EATS *ORDER', rules)).toBe('cat-food');
  });

  it('breaks a priority tie with the longer (more specific) keyword', () => {
    const rules = [
      rule({ keyword: 'uber', categoryId: 'cat-transport', priority: 1 }),
      rule({ keyword: 'uber eats', categoryId: 'cat-food', priority: 1 }),
    ];
    expect(suggestCategory('UBER EATS *ORDER', rules)).toBe('cat-food');
  });

  it('matches a keyword appearing anywhere in the description, not just at word boundaries', () => {
    const rules = [rule({ keyword: 'ber', categoryId: 'cat-transport' })];
    expect(suggestCategory('Uber Trip', rules)).toBe('cat-transport');
  });

  it('ignores rules that do not match at all, even with higher priority', () => {
    const rules = [
      rule({ keyword: 'padaria', categoryId: 'cat-food', priority: 100 }),
      rule({ keyword: 'uber', categoryId: 'cat-transport', priority: 1 }),
    ];
    expect(suggestCategory('UBER *TRIP', rules)).toBe('cat-transport');
  });
});

File: flutter_app/test/categorize_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:centavoo/categorize.dart';
import 'package:centavoo/models/category_rule.dart';

CategoryRule rule(String keyword, String categoryId, {int priority = 0}) {
  return CategoryRule(keyword: keyword, categoryId: categoryId, priority: priority);
}

void main() {
  test('returns the category for a keyword found in the description, case-insensitively', () {
    final rules = [rule('uber', 'transport')];
    expect(suggestCategory('Corrida de Uber até o hotel', rules), 'transport');
  });

  test('returns null when no keyword matches', () {
    final rules = [rule('uber', 'transport')];
    expect(suggestCategory('Jantar no restaurante', rules), isNull);
  });

  test('prefers the higher-priority rule when multiple keywords match', () {
    final rules = [rule('uber', 'transport', priority: 1), rule('uber eats', 'food', priority: 2)];
    expect(suggestCategory('Uber Eats - pedido', rules), 'food');
  });

  test('prefers the longer keyword when priorities tie', () {
    final rules = [rule('uber', 'transport'), rule('uber eats', 'food')];
    expect(suggestCategory('Uber Eats - pedido', rules), 'food');
  });
}
